pkg/controller/zerocapacity/zero_controller.go

```diff
@@ -0,0 +1,470 @@
+	namespace := request.Namespace
+
+	reqLogger := z.Log.WithValues("Request.Namespace", namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling " + resource)
+	defer reqLogger.Info("----- End Reconciling " + resource)
+
+	instance, err := reconciler.ResourceInstance(request.NamespacedName, z)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch %s CR '%s': %w", resource, request.Name, err)
+	}
+
+	phase := instance.Phase()
+	switch phase {
+	case "":
+		return reconcile.Result{}, instance.UpdatePhase(ZeroInitializing)
+	case ZeroInitializing:
+		return z.initializeResources(request, instance)
+	case ZeroSucceeded, ZeroFailed:
+		return z.cleanupResources(request)
+	}
+
+	infinispan := &v1.Infinispan{}
+	clusterName := instance.Cluster()
+	clusterObjKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+	if err := z.Get(context.Background(), clusterObjKey, infinispan); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("CR '%s' not found: %w", clusterName, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch CR '%s': %w", clusterName, err)
+	}
+
+	user := consts.DefaultOperatorUser
+	pass, err := users.PasswordFromSecret(user, infinispan.GetSecretName(), namespace, z.Kube)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	httpConfig := ispnclient.HttpConfig{
+		Username:  user,
+		Password:  pass,
+		Namespace: namespace,
+		Protocol:  strings.ToLower(string(infinispan.GetEndpointScheme())),
+	}
+	httpClient := curl.New(httpConfig, z.Kube)
+
+	if phase == ZeroInitialized {
+		return z.execute(httpClient, request, instance)
+	}
+	// Phase must be ZeroRunning, so wait for execution to complete
+	return z.waitForExecutionToComplete(httpClient, request, instance)
+}
+
+func (z *Controller) initializeResources(request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	ctx := context.Background()
+	name := request.Name
+	namespace := request.Namespace
+	clusterName := instance.Cluster()
+	clusterKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+
+	infinispan := &v1.Infinispan{}
+	if err := z.Client.Get(ctx, clusterKey, infinispan); err != nil {
+		retErr := fmt.Errorf("Unable to load Infinispan Cluster '%s': %w", clusterName, err)
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, retErr
+		}
+		return reconcile.Result{}, retErr
+	}
+
+	if err := ensureClusterStability(infinispan); err != nil {
```

Comment:
It seems to this condition is not checking correctly, as backup created directly after Infinispan CR reconciles with following error, not `Infinispan not stable` but it that time StatefulSet already present. Please see logs
```
{"level":"info","ts":1602694467.852345,"logger":"controller_infinispan","msg":"Creating a new StatefulSet","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan","StatefulSet.Name":"example-infinispan"}
{"level":"info","ts":1602694467.9548452,"logger":"controller_infinispan","msg":"End of the StetefulSet creation","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan"}
{"level":"info","ts":1602694468.1983776,"logger":"controller_infinispan","msg":"Pods IPs are not ready yet","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan"}
{"level":"info","ts":1602694468.1984246,"logger":"controller_infinispan","msg":"----- End Reconciling Infinispan.","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan"}
{"level":"info","ts":1602694468.1985173,"logger":"controller_infinispan","msg":"+++++ Reconciling Infinispan. Operator Version: 2.1.0.Dev01-28-gd31330f","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan"}
{"level":"info","ts":1602694468.2626894,"logger":"controller_infinispan","msg":"Pods IPs are not ready yet","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan"}
{"level":"info","ts":1602694468.2627418,"logger":"controller_infinispan","msg":"----- End Reconciling Infinispan.","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan"}
{"level":"info","ts":1602694497.8387942,"logger":"backup-controller","msg":"Reconciling ","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan-backup"}
{"level":"info","ts":1602694497.9161143,"logger":"backup-controller","msg":"----- End Reconciling ","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan-backup"}
{"level":"info","ts":1602694497.916394,"logger":"backup-controller","msg":"Reconciling ","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan-backup"}
{"level":"info","ts":1602694497.9174726,"logger":"backup-controller","msg":"----- End Reconciling ","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan-backup"}
{"level":"error","ts":1602694497.917662,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"backup-controller","request":"infinispan-operator/example-infinispan-backup","error":"Unable to load Infinispan StatefulSet 'example-infinispan': StatefulSet.apps \"\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/dvolodin/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1602694498.9179876,"logger":"backup-controller","msg":"Reconciling ","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan-backup"}
{"level":"info","ts":1602694498.9182673,"logger":"backup-controller","msg":"----- End Reconciling ","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan-backup"}
{"level":"error","ts":1602694498.9183495,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"backup-controller","request":"infinispan-operator/example-infinispan-backup","error":"Unable to load Infinispan StatefulSet 'example-infinispan': StatefulSet.apps \"\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/dvolodin/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1602694499.9187157,"logger":"backup-controller","msg":"Reconciling ","Request.Namespace":"infinispan-operator","Request.Name":"example-infinispan-backup"}
```